drivers/ethos/ethos.c

```diff
@@ -81,54 +81,51 @@ static void _reset_state(ethos_t *dev)
 {
     dev->state = WAIT_FRAMESTART;
     dev->frametype = 0;
-    dev->framesize = 0;
-    dev->accept_new = true;
 }
 
 static void _handle_char(ethos_t *dev, char c)
 {
     switch (dev->frametype) {
         case ETHOS_FRAME_TYPE_DATA:
-        case ETHOS_FRAME_TYPE_HELLO:
-        case ETHOS_FRAME_TYPE_HELLO_REPLY:
-            if (dev->accept_new) {
-                if (tsrb_add_one(&dev->inbuf, c) == 0) {
-                    dev->framesize++;
-                }
-            }
-            else {
+            if (tsrb_add_one(&dev->inbuf, c) < 0) {
                 //puts("lost frame");
                 dev->inbuf.reads = 0;
                 dev->inbuf.writes = 0;
-                _reset_state(dev);
+                dev->frametype = 0;
             }
             break;
-#ifdef MODULE_STDIO_ETHOS
         case ETHOS_FRAME_TYPE_TEXT:
-            dev->framesize++;
-            isrpipe_write_one(&stdio_uart_isrpipe, c);
+#ifdef MODULE_STDIO_ETHOS
+            if (isrpipe_write_one(&ethos_stdio_isrpipe, c) < 0) {
+                //puts("lost frame");
+                dev->inbuf.reads = 0;
+                dev->inbuf.writes = 0;
```

Comment:
This is the location where I should have started the thread on what losing bytes on the stdio byte would do to escaping.

Reply:
I hardened the error handling in the following way:

- The previous `frametype` determines which buffer (`tsrb` or `isrpipe`) is operated on.
- The buffer is reset (both `reads` and `writes` are set to 0)
- A `ETHOS_FRAME_DELIMITER` is added to the now reset buffer (as the buffer was now reset it will even work if the error was caused by an overflown buffer), to tell the handler thread to stop reading and just truncate the frame as read so far.